Support de S3

MAXOUXAX opened the pull request:
# Description
Ajout du support de S3 pour stocker les fichiers

# Modifications à noter 

- [x] Ajout d'un système de stratégie `FileSystemStorage` / `S3Storage` implémentant une classe commune
- [x] Gestion des métriques via un fichier `metrics.json` (requis car il sera très coûteux de lister tous les fichiers, leur type et leur taille à chaque fois que les métriques sont demandées)

# Performances

## Cache des fichiers redimensionnés `/{filename}?w={width}&h={height}`

Les images redimensionnées sont stockées dans le dossier `cache`, comme précédemment.
Afin de réduire la latence, on télécharge le fichier depuis S3 **seulement** si ce cache ne contient pas l'image redimensionnée

MAXOUXAX committed:
chore: added .gitattributes for .sh files

This is to prevent Git from changing EOL sequence to CRLF on Windows, causing apt to install nginx\r instead of nginx, leading to an error

MAXOUXAX committed:
feat: added support for all files type

This change added two new settings. The first one is a list of MIME types that are resizable (on which the w and h query parameter will apply) and the second one is the list of allowed MIME file types to be uploaded. If an user tries to upload a file, and the MIME type of this file isn't in the allowed MIME types list, an error will be raised

MAXOUXAX committed:
docs: added explanation on settings values

MAXOUXAX committed:
docs: preparing settings for S3 support

MAXOUXAX committed:
feat: added S3 support (still WIP)

MAXOUXAX committed:
refactor: metrics are storage dependant

MAXOUXAX committed:
fix: file is read twice leading to second read being empty because buffer has been consumed

MAXOUXAX committed:
chore: improved Dockerfile

MAXOUXAX committed:
docs: added section inside settings

MAXOUXAX committed:
chore: added back empty last lines

MAXOUXAX committed:
chore: missing new line in requirements.txt

MAXOUXAX committed:
fix: properly uploading and deleting temporary files when converting the image

MAXOUXAX committed:
chore: updated config according to new way to determine storage provider

MAXOUXAX committed:
feat: S3Storage provider now uses metrics.json to store metrics

MAXOUXAX committed:
feat: rebuilding metrics from S3 bucket

MAXOUXAX committed:
chore: added metrics-rebuilder service inside docker-compose

MAXOUXAX committed:
refactor: improved logging and added new metrics for execution time and date

MAXOUXAX committed:
refactor: updated storage methods to return metrics properly from the JSON file

MAXOUXAX committed:
feat: improved logging

MAXOUXAX committed:
chore: loading env variables from .env file for dev context

MAXOUXAX committed:
refactor: using /tmp instead of /temporary

MAXOUXAX committed:
fix: python buffer preventing logs from showing up

MAXOUXAX committed:
fix: removed decimals from file size

MAXOUXAX committed:
chore: removed unused variables

MAXOUXAX committed:
refactor: improved logging and robustness of metrics rebuilder

MAXOUXAX committed:
fix: handle case where S3_ENDPOINT is an empty string

MAXOUXAX committed:
chore: ignore metrics.json file

MAXOUXAX committed:
chore: properly comment method

MAXOUXAX committed:
refactor: improve file serving and caching functionality

MAXOUXAX committed:
fix: make sure imgpush.sock is deleted from image

MAXOUXAX committed:
refactor: store metrics file elsewhere than app folder to allow mounting it using a docker volume

MAXOUXAX committed:
chore: pushed .env (tokens are rotated :p)

MAXOUXAX committed:
fix: logging

MAXOUXAX committed:
chore: removed useless annotation

MAXOUXAX committed:
docs: added example environment variables file and updated README

MAXOUXAX committed:
docs: added AWS import script

MAXOUXAX committed:
refactor: put whole method in try catch to silently fail if an error occurs during metrics update

archibald-picq commented on .env.example:
```diff
@@ -0,0 +1,5 @@
+S3_ACCESS_KEY_ID=MY_S3_ACCESS_KEY_ID
+S3_SECRET_ACCESS_KEY=MY_S3_SECRET_ACCESS_KEY
+S3_BUCKET_NAME=MyBucket
+THREAD_COUNT=16
```
n'oublie pas le LF final

archibald-picq commented on .gitignore:
```diff
@@ -3,3 +3,5 @@ cache/
 *.pyc
 venv
+.env
+*metrics.json
```
n'oublie pas le LF final

MAXOUXAX committed:
chore: added missing EOF line

MAXOUXAX committed:
docs: using WSL to run bash commands

MAXOUXAX committed:
fix: file given to the storage provider before saving should be a BufferedReader

MAXOUXAX committed:
fix: add metrics volume to both containers inside docker compose file

MAXOUXAX committed:
docs: added statistics on metrics rebuilder based on the number of threads

MAXOUXAX committed:
refactor: improve logging inside main app

MAXOUXAX committed:
refactor: update metrics rebuilder to use environment variable for thread count

MAXOUXAX committed:
refactor: reformat code

MAXOUXAX committed:
docs: added WIP documentation

MAXOUXAX committed:
docs: update Setup guide to include S3 and filesystem configurations and configurable environment variables

MAXOUXAX committed:
chore: improve logs

MAXOUXAX committed:
feat: added support for specifying a directory to use inside the S3 bucket

MAXOUXAX committed:
docs: added thread count for metrics rebuilder inside setup process

MAXOUXAX committed:
feat: added failfast mechanism that checks the configuration before starting the webserver

MAXOUXAX committed:
docs: fix typos in stories

MAXOUXAX committed:
docs: improve docs on file importing

MAXOUXAX committed:
chore: precisly explaining the role of the method

MAXOUXAX committed:
docs: renamed all instances of image to file

MAXOUXAX committed:
chore: ignore sock files

MAXOUXAX committed:
docs: improve stories and setup

MAXOUXAX committed:
refactor: infer mime type from file extension

MAXOUXAX committed:
refactor: renamed fail fast python script and improved logging

MAXOUXAX committed:
docs: improve documentation

MAXOUXAX committed:
refactor: remove all mention of thread count since we're now infering mime type from file extension

MAXOUXAX committed:
docs: added newline and emojis for setup guides

MAXOUXAX committed:
fix: example env file using wrong format

MAXOUXAX committed:
docs: add logs section and add OVH as provider

MAXOUXAX committed:
chore: remove typo and simplify docker-compose.yml

MAXOUXAX committed:
docs: moved developer guide inside its own file

MAXOUXAX committed:
chore: force configuration of FILES_DIR env var

MAXOUXAX committed:
refactor: improve directory validation in FileSystemStorage

MAXOUXAX committed:
refactor: improve mime type validation and error handling in the fail fast mechanism

We check that the environment variables about mimetypes are in the valid format, and that every mimetype in them exists

MAXOUXAX committed:
docs: added troubleshooting guide listing all possible errors

MAXOUXAX committed:
fix: error when env variable not set

MAXOUXAX committed:
docs: added MIGRATION.md guide

MAXOUXAX committed:
fix: not failing if RESIZABLE_MIME_FILE_TYPES isn't set

Since the default value is plenty enough, we don't have to fail if no environment variable specifies the resizable mime types

MAXOUXAX committed:
refactor: storing date as UNIX timestamp

MAXOUXAX committed:
chore: updated metrics to include in-bucket folder name

MAXOUXAX committed:
feat: added /info route with version

MAXOUXAX committed:
docs: added guide to version imgpush

MAXOUXAX committed:
docs: updated stories with WIP information

MAXOUXAX committed:
feat: add metrics local stack

MAXOUXAX committed:
chore: rename clients in Prometheus tags

MAXOUXAX committed:
docs: added load testing documentation

MAXOUXAX committed:
feat: properly exported Grafana dashboard

MAXOUXAX committed:
chore: added directories inside gitignore

MAXOUXAX committed:
feat: updated Grafana dashboard

MAXOUXAX committed:
docs: improved documentation

MAXOUXAX committed:
docs: updated stories

MAXOUXAX committed:
docs: added CHANGELOG

MAXOUXAX committed:
chore: removed overrides in Grafana dashboard

MAXOUXAX committed:
feat: change alerts rule to reduce false positives

Commit:
fix import

Commit:
ouverture du fichier safe

Commit:
c'est une error et pas une info

Commit:
gestion fine des exceptions

Commit:
on evite les race conditions lors de l'update metrics

Commit:
on remplace le fichier temporaire par un fichier temporaire sécurisé

Commit:
ajout du support pour le dossier de destination

Commit:
Merge branch 'master' into feat/s3-support
